sdk/identity/azure-identity/src/private/token_cache_internals.hpp

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ * @brief Token cache internals and test hooks.
+ */
+
+#pragma once
+
+#include "token_cache.hpp"
+
+#include <azure/core/credentials/credentials.hpp>
+
+#include <chrono>
+#include <functional>
+#include <map>
+#include <shared_mutex>
+#include <string>
+#include <vector>
+
+namespace Azure { namespace Identity { namespace _detail {
+  /**
+   * @brief Implements internal aspects of token cache and provides test hooks.
+   *
+   */
+  class TokenCache::Internals final {
+  private:
+    Internals() = delete;
+    ~Internals() = delete;
+
+  public:
+    /**
+     * @brief Represents a unique set of characteristics that are used to distinguish between cache
+     * entries.
+     *
+     */
+    struct CacheKey
+    {
+      std::string TenantId; ///< Tenant ID.
+      std::string ClientId; ///< Client ID.
+      std::string AuthorityHost; ///< Authority Host.
+      std::string Scopes; ///< Authentication Scopes as a single string.
+
+      bool operator<(TokenCache::Internals::CacheKey const& other) const;
+    };
+
+    /**
+     * @brief The cache itself.
+     *
+     */
+    static std::map<CacheKey, Core::Credentials::AccessToken> Cache;
+
+    /**
+     * @brief Vector of expiration for the cache entries, from earliest to the latest.
+     *
+     */
+    static std::vector<std::pair<decltype(Core::Credentials::AccessToken::ExpiresOn), CacheKey>>
```

Comment:
I'm not sure we need this vector.

In most scenarios, there aren't too many distinct "Key" in a program. It doesn't matter if we just keep the cached entries forever in the map defined at L52.

Reply:
Removing this vector will also simplify the logic of `TokenCache::GetToken()`.

Reply:
I updated the code mostly as you suggested. I'll summarize in a separate message.
I would prefer the cache to be cleaned up from time to time, but it is logistically harder now. But I agree it probably will never be a problem. The benefits we get in exchange (concurrent read access to one token while other threads update the other token) are greater. I was thinking about it, it should still be possible to clean up cache, if we find it is a problem. It would be a bit more expensive now, since we would need to hold the lock for the `map` and try (without waiting) to get a write lock for the expired elements, I think it could be done, especially if we tell it to run for every 50th insertion, for example (or we can go exponential, run after the cache reaches `32`, `64`, `128`, ... in size so that we clean up proportionally often to the cache caliber).